core/math/math_funcs.h

```diff
@@ -37,245 +37,197 @@
+	return std::ceil(x);
 }
 
-_ALWAYS_INLINE_ double log1p(double p_x) {
-	return std::log1p(p_x);
-}
-_ALWAYS_INLINE_ float log1p(float p_x) {
-	return std::log1p(p_x);
+template <std::floating_point T>
+_ALWAYS_INLINE_ constexpr T pow(T x, T y) {
+	// Requires both params to be the same floating-type.
+	// Cast types manually when needed.
+	return std::pow(x, y);
 }
 
-_ALWAYS_INLINE_ double log2(double p_x) {
-	return std::log2(p_x);
-}
-_ALWAYS_INLINE_ float log2(float p_x) {
-	return std::log2(p_x);
+template <std::floating_point T>
+_ALWAYS_INLINE_ const T log(T x) {
+	return std::log(x);
 }
 
-_ALWAYS_INLINE_ double exp(double p_x) {
-	return std::exp(p_x);
-}
-_ALWAYS_INLINE_ float exp(float p_x) {
-	return std::exp(p_x);
+template <std::floating_point T>
+_ALWAYS_INLINE_ constexpr T log1p(T x) {
+	return std::log1p(x);
 }
 
-_ALWAYS_INLINE_ bool is_nan(double p_val) {
-	return std::isnan(p_val);
+template <std::floating_point T>
+_ALWAYS_INLINE_ constexpr T log2(T x) {
+	return std::log2(x);
 }
 
-_ALWAYS_INLINE_ bool is_nan(float p_val) {
-	return std::isnan(p_val);
+template <std::floating_point T>
+_ALWAYS_INLINE_ constexpr T exp(T x) {
+	return std::exp(x);
 }
 
-_ALWAYS_INLINE_ bool is_inf(double p_val) {
-	return std::isinf(p_val);
+template <std::floating_point T>
+_ALWAYS_INLINE_ constexpr bool is_nan(T val) {
+	return std::isnan(val);
 }
 
-_ALWAYS_INLINE_ bool is_inf(float p_val) {
-	return std::isinf(p_val);
+template <std::floating_point T>
+_ALWAYS_INLINE_ constexpr bool is_inf(T val) {
+	return std::isinf(val);
 }
 
-// These methods assume (p_num + p_den) doesn't overflow.
-_ALWAYS_INLINE_ int32_t division_round_up(int32_t p_num, int32_t p_den) {
-	int32_t offset = (p_num < 0 && p_den < 0) ? 1 : -1;
-	return (p_num + p_den + offset) / p_den;
-}
-_ALWAYS_INLINE_ uint32_t division_round_up(uint32_t p_num, uint32_t p_den) {
-	return (p_num + p_den - 1) / p_den;
-}
-_ALWAYS_INLINE_ int64_t division_round_up(int64_t p_num, int64_t p_den) {
-	int32_t offset = (p_num < 0 && p_den < 0) ? 1 : -1;
-	return (p_num + p_den + offset) / p_den;
-}
-_ALWAYS_INLINE_ uint64_t division_round_up(uint64_t p_num, uint64_t p_den) {
-	return (p_num + p_den - 1) / p_den;
+template <std::integral T>
+_ALWAYS_INLINE_ constexpr T division_round_up(T numerator, T denominator) {
+	// WARNING: Assumes numerator + denominator - 1 will not overflow.
+	if constexpr (std::is_unsigned_v<T>) {
+		return (numerator + denominator - 1) / denominator;
+	} else {
+		// Avoid the overflow by detecting sign of both ops.
+		if ((numerator ^ denominator) >= 0) {
+			// Round up.
+			return (numerator + denominator - 1) / denominator;
+		} else {
+			// Do not round up.
+			return numerator / denominator;
+		}
+	}
 }
 
-_ALWAYS_INLINE_ bool is_finite(double p_val) {
-	return std::isfinite(p_val);
-}
-_ALWAYS_INLINE_ bool is_finite(float p_val) {
-	return std::isfinite(p_val);
+template <std::floating_point T>
+_ALWAYS_INLINE_ bool is_finite(T val) {
```

Comment:
```suggestion
_ALWAYS_INLINE_ constexpr bool is_finite(T val) {
```
std::isfinite() is constexpr